drivers/flash/flash_stm32f2x.c

```diff
@@ -0,0 +1,169 @@
+/*
+ * Copyright (c) 2021 Linaro Limited
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <kernel.h>
+#include <device.h>
+#include <string.h>
+#include <drivers/flash.h>
+#include <init.h>
+#include <soc.h>
+
+#include "flash_stm32.h"
+
+#define STM32F2X_SECTOR_MASK		((uint32_t) 0xFFFFFF07)
+
+bool flash_stm32_valid_range(const struct device *dev, off_t offset,
+			     uint32_t len,
+			     bool write)
+{
+	ARG_UNUSED(write);
+
+	return flash_stm32_range_exists(dev, offset, len);
+}
+
+static int write_byte(const struct device *dev, off_t offset, uint8_t val)
+{
+	FLASH_TypeDef *regs = FLASH_STM32_REGS(dev);
+	uint32_t tmp;
+	int rc;
+
+	/* if the control register is locked, do not fail silently */
+	if (regs->CR & FLASH_CR_LOCK) {
+		return -EIO;
+	}
+
+	rc = flash_stm32_wait_flash_idle(dev);
+	if (rc < 0) {
+		return rc;
+	}
+
+	regs->CR &= CR_PSIZE_MASK;
+	regs->CR |= FLASH_PSIZE_BYTE;
+	regs->CR |= FLASH_CR_PG;
+
+	/* flush the register write */
+	tmp = regs->CR;
+
+	*((uint8_t *) offset + CONFIG_FLASH_BASE_ADDRESS) = val;
+
+	rc = flash_stm32_wait_flash_idle(dev);
+	regs->CR &= (~FLASH_CR_PG);
+
+	return rc;
+}
+
+static int erase_sector(const struct device *dev, uint32_t sector)
+{
+	FLASH_TypeDef *regs = FLASH_STM32_REGS(dev);
+	uint32_t tmp;
+	int rc;
+
+	/* if the control register is locked, do not fail silently */
+	if (regs->CR & FLASH_CR_LOCK) {
+		return -EIO;
+	}
+
+	rc = flash_stm32_wait_flash_idle(dev);
+	if (rc < 0) {
+		return rc;
+	}
+
+	regs->CR &= STM32F2X_SECTOR_MASK;
+	regs->CR |= FLASH_CR_SER | (sector << 3);
+	regs->CR |= FLASH_CR_STRT;
+
+	/* flush the register write */
+	tmp = regs->CR;
+
+	rc = flash_stm32_wait_flash_idle(dev);
+	regs->CR &= ~(FLASH_CR_SER | FLASH_CR_SNB);
+
+	return rc;
+}
+
+int flash_stm32_block_erase_loop(const struct device *dev,
+				 unsigned int offset,
+				 unsigned int len)
+{
+	struct flash_pages_info info;
+	uint32_t start_sector, end_sector;
+	uint32_t i;
+	int rc = 0;
+
+	rc = flash_get_page_info_by_offs(dev, offset, &info);
+	if (rc) {
+		return rc;
+	}
+	start_sector = info.index;
+	rc = flash_get_page_info_by_offs(dev, offset + len - 1, &info);
+	if (rc) {
+		return rc;
+	}
+	end_sector = info.index;
+
+	for (i = start_sector; i <= end_sector; i++) {
+		rc = erase_sector(dev, i);
+		if (rc < 0) {
+			break;
+		}
+	}
+
+	return rc;
+}
+
+int flash_stm32_write_range(const struct device *dev, unsigned int offset,
+			    const void *data, unsigned int len)
+{
+	int i, rc = 0;
+
+	for (i = 0; i < len; i++, offset++) {
+		rc = write_byte(dev, offset, ((const uint8_t *) data)[i]);
+		if (rc < 0) {
+			return rc;
+		}
+	}
+
+	return rc;
+}
+
+/*
+ * Different SoC flash layouts are specified in across various
+ * reference manuals, but the flash layout for a given number of
+ * sectors is consistent across these manuals, with one "gotcha". The
+ * number of sectors is given by the HAL as FLASH_SECTOR_TOTAL.
```

Comment:
From what I can see there is only one flash layout on F2 series. This section would require an update.